app/src/main/java/org/thoughtcrime/securesms/database/helpers/SessionStoreMigrationHelper.java

```diff
@@ -0,0 +1,103 @@
+package org.thoughtcrime.securesms.database.helpers;
+
+
+import android.content.ContentValues;
+import android.content.Context;
+
+import net.zetetic.database.sqlcipher.SQLiteDatabase;
+
+import org.signal.core.util.Conversions;
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.database.SessionDatabase;
+import org.whispersystems.libsignal.state.SessionRecord;
+import org.whispersystems.signalservice.api.push.SignalServiceAddress;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+
+class SessionStoreMigrationHelper {
+
+  private static final String TAG = Log.tag(SessionStoreMigrationHelper.class);
+
+  private static final String SESSIONS_DIRECTORY_V2 = "sessions-v2";
+  private static final Object FILE_LOCK             = new Object();
+
+  private static final int SINGLE_STATE_VERSION   = 1;
+  private static final int ARCHIVE_STATES_VERSION = 2;
+  private static final int PLAINTEXT_VERSION      = 3;
+  private static final int CURRENT_VERSION        = 3;
+
+  static void migrateSessions(Context context, SQLiteDatabase database) {
+    File directory = new File(context.getFilesDir(), SESSIONS_DIRECTORY_V2);
+
+    if (directory.exists()) {
+      File[] sessionFiles = directory.listFiles();
+
+      if (sessionFiles != null) {
+        for (File sessionFile : sessionFiles) {
+          try {
+            String[] parts   = sessionFile.getName().split("[.]");
+            String   address = parts[0];
+
+            int deviceId;
+
+            if (parts.length > 1) deviceId = Integer.parseInt(parts[1]);
+            else                  deviceId = SignalServiceAddress.DEFAULT_DEVICE_ID;
+
+            FileInputStream in            = new FileInputStream(sessionFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at lines 49, 55. The resource is closed at line 56. There are other execution paths that do not contain closure statements, for example, when an if condition at line 51 evaluates to true. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).